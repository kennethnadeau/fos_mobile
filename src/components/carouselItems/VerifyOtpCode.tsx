import React, {forwardRef} from 'react';
import CarouselItemContainer from './CarouselItemContainer';
import {Text} from 'react-native-elements';
import {s, vs} from 'react-native-size-matters';
import {Colors} from '@fos/themes';
import {useTranslation, Trans} from 'react-i18next';
import OTPInputView from '@twotalltotems/react-native-otp-input';
import {View, StyleSheet, TextStyle} from 'react-native';
import FooterText from './FooterText';

export type VerificationCodeStatus = 'unverified' | 'verified' | 'invalid';

export type VerifyCodeProps = {
  mobileNumber?: string;
  verificationStatus?: VerificationCodeStatus;
  code?: string;
  onCodeChange?: (code: string) => void;
  onCodeFilled?: (code: string) => void;
  onResendPress: () => void;
};

const VerifyOtpCode = forwardRef<OTPInputView, VerifyCodeProps>(
  (
    {
      mobileNumber,
      code,
      verificationStatus = 'unverified',
      onCodeChange,
      onCodeFilled,
      onResendPress,
    },
    ref,
  ) => {
    const {t} = useTranslation('carouselItems');

    const verificationStatusBorderColor: {
      [key in VerificationCodeStatus]: TextStyle;
    } = {
      unverified: styles.codeInputField,
      invalid: {...styles.codeInputField, ...styles.invalid},
      verified: {...styles.codeInputField, ...styles.verified},
    };

    return (
      <CarouselItemContainer
        containerProps={{
          accessible: true,
          accessibilityLabel: t('Enter Verification Code'),
        }}
        containerStyle={{
          paddingHorizontal: s(16),
        }}
        header={t('Enter Verification Code')}>
        <View accessibilityLabel={t('OTP code input')} accessible>
          <OTPInputView
            autoFocusOnLoad={false}
            code={code}
            codeInputFieldStyle={
              verificationStatusBorderColor[verificationStatus]
            }
            onCodeChanged={onCodeChange}
            onCodeFilled={onCodeFilled}
            pinCount={6}
            ref={ref}
            selectionColor={Colors.secondary}
            style={styles.otpInput}
          />
        </View>

        <FooterText style={styles.sentToDescription}>
          <Trans t={t} values={{mobileNumber}}>
            We texted a verification code
          </Trans>
        </FooterText>
        <FooterText style={styles.resend}>
          <Trans
            components={{
              Text: <Text onPress={onResendPress} />,
            }}
            t={t}>
            Resend Code
          </Trans>
        </FooterText>
      </CarouselItemContainer>
    );
  },
);

const styles = StyleSheet.create({
  codeInputField: {
    borderRadius: s(8),
    fontSize: s(32),
    fontWeight: '900',
    height: s(60),
  },
  invalid: {
    borderColor: Colors.error,
  },
  otpInput: {
    height: vs(100),
    paddingHorizontal: s(16),
  },
  resend: {
    marginTop: vs(50),
    textAlign: 'center',
  },
  sentToDescription: {
    marginTop: vs(20),
    paddingHorizontal: s(30),
    textAlign: 'center',
  },
  verified: {
    borderColor: Colors.success,
  },
});

export default VerifyOtpCode;
