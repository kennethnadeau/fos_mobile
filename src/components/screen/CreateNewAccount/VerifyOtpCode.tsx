import React, {FC} from 'react';
import CarouselItemContainer from './CarouselItemContainer';
import {Text} from 'react-native-elements';
import {s, vs} from 'react-native-size-matters';
import {Colors} from '@fos/themes';
import {useTranslation, Trans} from 'react-i18next';
import OTPInputView from '@twotalltotems/react-native-otp-input';
import {View, StyleSheet, TextStyle} from 'react-native';
import FooterText from './FooterText';

export type VerificationCodeStatus = 'unverified' | 'verified' | 'invalid';

export type VerifyCodeProps = {
  mobileNumber?: string;
  verificationStatus?: VerificationCodeStatus;
  code?: string;
  onCodeChange?: (code: string) => void;
  onCodeFilled?: (code: string) => void;
  onResendPress: () => void;
};

const VerifyOtpCode: FC<VerifyCodeProps> = ({
  mobileNumber,
  code,
  verificationStatus = 'unverified',
  onCodeChange,
  onCodeFilled,
  onResendPress,
}) => {
  const {t: translate} = useTranslation();

  const t = (path: string) => translate(`screens.createNewAccount.${path}`);
  const verificationStatusBorderColor: {
    [key in VerificationCodeStatus]: TextStyle;
  } = {
    unverified: styles.codeInputField,
    invalid: {...styles.codeInputField, ...styles.invalid},
    verified: {...styles.codeInputField, ...styles.verified},
  };

  return (
    <CarouselItemContainer
      containerProps={{
        accessible: true,
        accessibilityLabel: t('verifyOtpCodeHeader'),
      }}
      containerStyle={{
        paddingHorizontal: s(16),
      }}
      header={t('verifyOtpCodeHeader')}>
      <View
        accessibilityLabel={t('verifyOtpCodeInput.otpCodeInput')}
        accessible>
        <OTPInputView
          autoFocusOnLoad={false}
          code={code}
          codeInputFieldStyle={
            verificationStatusBorderColor[verificationStatus]
          }
          onCodeChanged={onCodeChange}
          onCodeFilled={onCodeFilled}
          pinCount={6}
          selectionColor={Colors.secondary}
          style={styles.otpInput}
        />
      </View>

      <FooterText style={styles.sentToDescription}>
        <Trans
          i18nKey="screens.createNewAccount.verifyOtpCodeFooter.sentToMobileNumber"
          values={{mobileNumber}}
        />
      </FooterText>
      <FooterText style={styles.resend}>
        <Trans
          components={{
            Text: <Text onPress={onResendPress} />,
          }}
          i18nKey="screens.createNewAccount.verifyOtpCodeFooter.resendCode"
        />
      </FooterText>
    </CarouselItemContainer>
  );
};

const styles = StyleSheet.create({
  codeInputField: {
    borderRadius: s(8),
    fontSize: s(32),
    fontWeight: '900',
    height: s(60),
  },
  invalid: {
    borderColor: Colors.error,
  },
  otpInput: {
    height: vs(100),
    paddingHorizontal: s(16),
  },
  resend: {
    marginTop: vs(50),
    textAlign: 'center',
  },
  sentToDescription: {
    marginTop: vs(20),
    paddingHorizontal: s(30),
    textAlign: 'center',
  },
  verified: {
    borderColor: Colors.success,
  },
});

export default VerifyOtpCode;
