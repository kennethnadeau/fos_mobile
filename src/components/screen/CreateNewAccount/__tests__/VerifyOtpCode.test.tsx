import React from 'react';
import {render, RenderAPI} from '@fos/utils/testing';
import VerifyOtpCode from '../VerifyOtpCode';
import enUS from '@fos/locale/languages/en-US';

const localization = enUS.translation.screens.createNewAccount;

describe('<VerifyOtpCode />', () => {
  let verifyOtpCode: RenderAPI;

  beforeEach(() => {
    verifyOtpCode = render(
      <VerifyOtpCode mobileNumber="123456789" onResendPress={jest.fn()} />,
    );
  });

  it('should have carousel item description', () => {
    const actual = verifyOtpCode.queryByA11yLabel(
      localization.verifyOtpCodeHeader,
    );

    expect(actual).not.toBeNull();
  });

  it('should show header', () => {
    const actual = verifyOtpCode.queryByText(localization.verifyOtpCodeHeader);

    expect(actual).not.toBeNull();
  });

  it('should show OTP code input', () => {
    const actual = verifyOtpCode.queryByA11yLabel(
      localization.verifyOtpCodeInput.otpCodeInput,
    );

    expect(actual).not.toBeNull();
    expect(actual?.children).toHaveLength(1);
  });

  it('should match snapshot', () => {
    expect(verifyOtpCode.toJSON()).toMatchSnapshot();
  });
});
