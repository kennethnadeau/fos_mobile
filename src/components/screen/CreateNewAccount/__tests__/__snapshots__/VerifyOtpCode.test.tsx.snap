// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VerifyOtpCode /> should match snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    accessibilityLabel="Enter Verification Code"
    accessible={true}
    style={
      Array [
        Object {
          "alignItems": "center",
          "flex": 1,
          "paddingHorizontal": 85.71428571428571,
        },
        Object {
          "paddingHorizontal": 34.285714285714285,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 27.5,
          "marginVertical": 39.23529411764706,
        }
      }
      theme={
        Object {
          "Button": Object {
            "buttonStyle": Array [
              Object {
                "backgroundColor": "#4CA1D6",
              },
            ],
            "titleStyle": Array [
              Object {
                "color": "white",
                "fontSize": 34.285714285714285,
              },
            ],
          },
          "Icon": Object {
            "color": "#2089dc",
          },
          "Input": Object {
            "inputContainerStyle": Object {
              "alignItems": "center",
              "borderBottomWidth": 1.4999999999999998,
              "borderColor": "#fff",
              "borderRadius": 64.28571428571428,
              "borderWidth": 1.4999999999999998,
              "overflow": "hidden",
              "padding": 9.428571428571429,
            },
            "inputStyle": Object {
              "alignItems": "center",
              "color": "#fff",
              "padding": 9.428571428571429,
            },
            "placeholderTextColor": "#7E8395",
          },
          "Text": Object {
            "style": Object {
              "color": "white",
              "fontSize": 34.285714285714285,
            },
          },
          "colors": Object {
            "disabled": "hsl(208, 8%, 90%)",
            "divider": "#bcbbc1",
            "error": "#ff190c",
            "grey0": "#393e42",
            "grey1": "#43484d",
            "grey2": "#5e6977",
            "grey3": "#86939e",
            "grey4": "#bdc6cf",
            "grey5": "#e1e8ee",
            "greyOutline": "#bbb",
            "platform": Object {
              "android": Object {
                "error": "#f44336",
                "primary": "#2196f3",
                "secondary": "#9C27B0",
                "success": "#4caf50",
                "warning": "#ffeb3b",
              },
              "ios": Object {
                "error": "#ff3b30",
                "primary": "#007aff",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
            },
            "primary": "#2089dc",
            "searchBg": "#303337",
            "secondary": "#8F0CE8",
            "success": "#52c41a",
            "warning": "#faad14",
          },
        }
      }
    >
      Enter Verification Code
    </Text>
    <View
      accessibilityLabel="OTP code input"
      accessible={true}
    >
      <OtpInputViewMock
        autoFocusOnLoad={false}
        codeInputFieldStyle={
          Object {
            "borderRadius": 17.142857142857142,
            "fontSize": 68.57142857142857,
            "fontWeight": "900",
            "height": 128.57142857142856,
          }
        }
        pinCount={6}
        selectionColor="#4CA1D6"
        style={
          Object {
            "height": 196.17647058823528,
            "paddingHorizontal": 34.285714285714285,
          }
        }
      />
    </View>
    <Text
      style={
        Object {
          "color": "#7E8395",
          "fontSize": 34.285714285714285,
          "marginTop": 39.23529411764706,
          "paddingHorizontal": 64.28571428571428,
          "textAlign": "center",
        }
      }
      theme={
        Object {
          "Button": Object {
            "buttonStyle": Array [
              Object {
                "backgroundColor": "#4CA1D6",
              },
            ],
            "titleStyle": Array [
              Object {
                "color": "white",
                "fontSize": 34.285714285714285,
              },
            ],
          },
          "Icon": Object {
            "color": "#2089dc",
          },
          "Input": Object {
            "inputContainerStyle": Object {
              "alignItems": "center",
              "borderBottomWidth": 1.4999999999999998,
              "borderColor": "#fff",
              "borderRadius": 64.28571428571428,
              "borderWidth": 1.4999999999999998,
              "overflow": "hidden",
              "padding": 9.428571428571429,
            },
            "inputStyle": Object {
              "alignItems": "center",
              "color": "#fff",
              "padding": 9.428571428571429,
            },
            "placeholderTextColor": "#7E8395",
          },
          "Text": Object {
            "style": Object {
              "color": "white",
              "fontSize": 34.285714285714285,
            },
          },
          "colors": Object {
            "disabled": "hsl(208, 8%, 90%)",
            "divider": "#bcbbc1",
            "error": "#ff190c",
            "grey0": "#393e42",
            "grey1": "#43484d",
            "grey2": "#5e6977",
            "grey3": "#86939e",
            "grey4": "#bdc6cf",
            "grey5": "#e1e8ee",
            "greyOutline": "#bbb",
            "platform": Object {
              "android": Object {
                "error": "#f44336",
                "primary": "#2196f3",
                "secondary": "#9C27B0",
                "success": "#4caf50",
                "warning": "#ffeb3b",
              },
              "ios": Object {
                "error": "#ff3b30",
                "primary": "#007aff",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
            },
            "primary": "#2089dc",
            "searchBg": "#303337",
            "secondary": "#8F0CE8",
            "success": "#52c41a",
            "warning": "#faad14",
          },
        }
      }
    >
      We texted a verification code
    </Text>
    <Text
      style={
        Object {
          "color": "#7E8395",
          "fontSize": 34.285714285714285,
          "marginTop": 98.08823529411764,
          "textAlign": "center",
        }
      }
      theme={
        Object {
          "Button": Object {
            "buttonStyle": Array [
              Object {
                "backgroundColor": "#4CA1D6",
              },
            ],
            "titleStyle": Array [
              Object {
                "color": "white",
                "fontSize": 34.285714285714285,
              },
            ],
          },
          "Icon": Object {
            "color": "#2089dc",
          },
          "Input": Object {
            "inputContainerStyle": Object {
              "alignItems": "center",
              "borderBottomWidth": 1.4999999999999998,
              "borderColor": "#fff",
              "borderRadius": 64.28571428571428,
              "borderWidth": 1.4999999999999998,
              "overflow": "hidden",
              "padding": 9.428571428571429,
            },
            "inputStyle": Object {
              "alignItems": "center",
              "color": "#fff",
              "padding": 9.428571428571429,
            },
            "placeholderTextColor": "#7E8395",
          },
          "Text": Object {
            "style": Object {
              "color": "white",
              "fontSize": 34.285714285714285,
            },
          },
          "colors": Object {
            "disabled": "hsl(208, 8%, 90%)",
            "divider": "#bcbbc1",
            "error": "#ff190c",
            "grey0": "#393e42",
            "grey1": "#43484d",
            "grey2": "#5e6977",
            "grey3": "#86939e",
            "grey4": "#bdc6cf",
            "grey5": "#e1e8ee",
            "greyOutline": "#bbb",
            "platform": Object {
              "android": Object {
                "error": "#f44336",
                "primary": "#2196f3",
                "secondary": "#9C27B0",
                "success": "#4caf50",
                "warning": "#ffeb3b",
              },
              "ios": Object {
                "error": "#ff3b30",
                "primary": "#007aff",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
            },
            "primary": "#2089dc",
            "searchBg": "#303337",
            "secondary": "#8F0CE8",
            "success": "#52c41a",
            "warning": "#faad14",
          },
        }
      }
    >
      Resend Code
    </Text>
  </View>
</RNCSafeAreaProvider>
`;
